src/Makefile.am

```diff
@@ -9,7 +9,7 @@ lib_LTLIBRARIES = libica.la
 libica_la_CFLAGS = ${AM_CFLAGS} -I${srcdir}/include -I${srcdir}/../include \
 		   -fvisibility=hidden -pthread
 libica_la_CCASFLAGS = ${AM_CFLAGS}
-libica_la_LIBADD = @LIBS@ -lrt -lcrypto -ldl
+libica_la_LIBADD = @LIBS@ -lrt -lcrypto -ldl -lm
```

Comment:
Why do we link libmath ? As far as i can tell you only need it for "pow(2, x)" which you could easily do just using the shift operator "2 << x" ...

Reply:
Hm .. I like the idea and did some tests. To calculate pow(2,n) we would have to code (1ULL << n).
```
printf("  1 << 31 = %llu\n",(1ULL << 31));
printf("  1 << 32 = %llu\n",(1ULL << 32));
printf("  1 << 63 = %llu\n",(1ULL << 63));
printf("  1 << 64 = %llu\n",(1ULL << 64));
printf("  pow(2,64) = %f\n",pow(2,64));
```
The output is:
```
  1 << 31 = 2147483648
  1 << 32 = 4294967296
  1 << 63 = 9223372036854775808
  1 << 64 = 0
  pow(2,64) = 18446744073709551616.000000
```
So we have a problem if the ctr_width is 64. And with an AES counter the numbers can get even bigger as the AES block size is 16. What do you think?

Reply:
So IMO the check
```
data_length > pow(2,ctr_width)
```
is incorrect anyways: data_length are bytes not blocks, so data_length should be divided by block size (and ceil-ed if it is not a multiple of block size), since the counter is incremented per processed block:
```
(data_length + BLOCK_SIZE - 1) / BLOCK_SIZE > ...
```

Yes, the shift operator is limited, but the left hand side of above comparison can never exceed ULL max.
I think you should rethink the check. 

Also, the the counter input ctr may not be zero. If the function can be called in a streaming fashion, you have to consider that also.